password/src/Data/Password/Validate.hs

```diff
@@ -126,122 +143,139 @@ data CharacterCategory
   -- ^ ASCII digits
   deriving (Eq, Ord, Show)
 
--- | Convert 'CharacterCategory' into associated predicate function
+-- | Convert a 'CharacterCategory' into its associated predicate function
+--
+-- @since 2.1.0.0
 categoryToPredicate :: CharacterCategory -> (Char -> Bool)
 categoryToPredicate = \case
   Uppercase -> isAsciiUpper
   Lowercase -> isAsciiLower
   Special -> isSpecial
   Digit -> isDigit
 
--- | Possible reason of 'Password' being invalid
+-- | Possible reason for a 'Password' to be invalid.
+--
+-- @since 2.1.0.0
 data InvalidReason
   = PasswordTooShort !Int !Int
   -- ^ Length of 'Password' is too short.
   --
-  -- Expected at least 'Int' characters but the actual length is 'Int'
+  -- @PasswordTooShort expected found@
   | PasswordTooLong !Int !Int
   -- ^ Length of 'Password' is too long.
   --
-  -- Expected at maximum of 'Int' characters, but the actual length is 'Int'
-  | NotEnoughReqChars CharacterCategory !Int !Int
+  -- @PasswordTooLong expected found@
+  | NotEnoughReqChars !CharacterCategory !Int !Int
   -- ^ 'Password' does not contain required number of characters.
   --
-  -- Expected at least 'Int' characters of 'CharacterCategory' but the password only
-  -- contains 'Int'
+  -- @NotEnoughReqChars category expected found@
   | InvalidCharacters !Text
   -- ^ 'Password' contains characters that cannot be used
-  | InvalidPasswordPolicy [InvalidPolicyReason]
-  -- ^ 'PasswordPolicy' is invalid
   deriving (Eq, Ord, Show)
 
 -- | Possible reason of 'PasswordPolicy' being invalid
+--
+-- @since 2.1.0.0
 data InvalidPolicyReason
   = InvalidLength !Int !Int
   -- ^ Value of 'minimumLength' is bigger than 'maximumLength'
+  --
+  -- @InvalidLength min max@
   | MaxLengthBelowZero !Int
-  -- ^ Value of 'maximumLength' is less than zero
+  -- ^ Value of 'maximumLength' is zero or less
   | InvalidCharSetPredicate !CharacterCategory !Int
   -- ^ 'CharSetPredicate' does not return 'True' for a 'CharacterCategory' that
   -- requires at least 'Int' characters in the password
   deriving (Eq, Ord, Show)
 
+-- | Result of validating a 'Password'.
+--
+-- Note that if the 'PasswordPolicy' is invalid, this will never return 'ValidPassword'.
+--
+-- @since 2.1.0.0
+data ValidationResult
+  = ValidPassword
+  -- ^ The 'Password' conforms to the validation parameters
+  | InvalidPassword [InvalidReason]
+  -- ^ 'Password' failed to be validated for the given reasons
+  | InvalidPolicy [InvalidPolicyReason]
+  -- ^ 'PasswordPolicy' is invalid
+  deriving (Eq, Show)
 
--- | Check if given 'Password' fullfills all the Policies,
--- return true if given password is valid
+-- | Checks if the given 'Password' adheres to the given 'PasswordPolicy'
+-- and 'CharSetPredicate', and returns @True@ if given a valid password.
 --
--- This function is equivalent to @null $ validatePassword policy password@
+-- This function is equivalent to @validatePassword policy charSetPredicate password == ValidPassword@
```

Comment:
```suggestion
-- This function is equivalent to @'validatePassword' policy charSetPredicate password == 'ValidPassword'@
```

I think Haddock links like this work correctly.

Reply:
Yup, works.